cli/src/commands/extensions/permissions.rs

```diff
@@ -61,6 +62,74 @@ impl Permission {
         }
     }
+
+    pub fn subset_of(&self, parent: &Permission) -> Result<Permission> {
+        use Permission::*;
+        match (parent, self) {
+            // Child deny-all always succeeds, returning deny-all.
+            (_, &Boolean(false)) => Ok(Boolean(false)),
+            // Parent deny-all fails with all child permissions but deny-all.
+            (&Boolean(false), _) => Err(anyhow!("Requested permissions incompatible with parent")),
+            // Parent allow-all always succeeds, returning the child's permissions.
+            (&Boolean(true), child) => Ok(child.clone()),
+            // Child allow-all fails with more restrictive parent permissions.
+            (_, &Boolean(true)) => Err(anyhow!("Requested permissions incompatible with parent")),
+            // Parent set vs child set have to be validated.
+            // This will error if child is not subset of parent, and return the child set otherwise.
+            (&List(ref parent), &List(ref child)) => Permission::paths_subset(parent, child)
+                .map(|()| Permission::List(child.clone()))
+                .map_err(|mismatches| {
+                    anyhow!(
+                        "Requested permission paths incompatible with parent: {}",
+                        mismatches.join(", ")
+                    )
+                }),
+        }
+    }
+
+    fn paths_subset(parent: &[String], child: &[String]) -> StdResult<(), Vec<String>> {
+        // Let P be the set of all paths.
+        //
+        // Let "<" be the partial order relation over P such that "a < b" reads "a is
+        // prefix of b" where a, b are two paths in P.
+        //
+        // Let "<<" be the partial order relation over the power set of P, such that "A
+        // << B" reads "A is paths-subset of B" where A and B are subsets of P.
+        //
+        // A << B if and only if, for each a in A, there exist at least one b in B such
+        // that a < b.
+        //
+        // The above definition is tested in `tests::paths_subset_algorithm` below.
+        //
+        // In this method, A is `child` and B is `parent`.
+
+        let parent_paths = parent.iter().map(PathBuf::from).collect::<Vec<_>>();
+        let child_paths = child.iter().map(PathBuf::from).collect::<Vec<_>>();
+
+        // List all child paths for which no parent path exists such that child <
+        // parent. O(n*m) where n = child_paths,len() and m =
+        // parent_paths.len().
+        let without_parent: Vec<_> = child_paths
+            .iter()
+            .filter_map(|child| {
+                // Using PathBuf::starts_with rather than String::starts_with in order to get
+                // the correct semantics.
+                if !parent_paths.iter().any(|p| child.starts_with(p)) {
+                    Some(child.to_string_lossy().to_string())
+                } else {
+                    None
+                }
```

Comment:
It feels kinda weird to put the `None` before the `Some` branch, but I don't like the negation either. Wish we had a `none` method that's just `!any`, it would be perfect.